codec-dns/src/main/java/io/netty/handler/codec/dns/DnsMessageUtil.java

```diff
@@ -0,0 +1,179 @@
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package io.netty.handler.codec.dns;
+
+import io.netty.channel.AddressedEnvelope;
+import io.netty.util.internal.StringUtil;
+
+import java.net.SocketAddress;
+import java.util.List;
+
+/**
+ * Provides some utility methods for DNS message implementations.
+ */
+final class DnsMessageUtil {
+
+    static StringBuilder appendQuery(StringBuilder buf, DnsQuery query) {
+        appendQueryHeader(buf, query);
+        appendAllRecords(buf, query);
+        return buf;
+    }
+
+    static StringBuilder appendResponse(StringBuilder buf, DnsResponse response) {
+        appendResponseHeader(buf, response);
+        appendAllRecords(buf, response);
+        return buf;
+    }
+
+    private static void appendQueryHeader(StringBuilder buf, DnsQuery msg) {
+        buf.append(StringUtil.simpleClassName(msg));
+        buf.append('(');
+        appendAddresses(buf, msg);
+        buf.append(msg.id());
+        buf.append(", ");
+        buf.append(msg.opCode());
+        if (msg.isRecursionDesired()) {
+            buf.append(", RD");
+        }
+        if (msg.z() != 0) {
+            buf.append(", Z: ");
+            buf.append(msg.z());
+        }
+        buf.append(')');
+    }
+
+    private static void appendResponseHeader(StringBuilder buf, DnsResponse msg) {
+        buf.append(StringUtil.simpleClassName(msg));
+        buf.append('(');
+        appendAddresses(buf, msg);
+        buf.append(msg.id());
+        buf.append(", ");
+        buf.append(msg.opCode());
+        buf.append(", ");
+        buf.append(msg.code());
+        buf.append(',');
+
+        boolean hasComma = true;
+        if (msg.isRecursionDesired()) {
+            hasComma = false;
+            buf.append(" RD");
+        }
+        if (msg.isAuthoritativeAnswer()) {
+            hasComma = false;
+            buf.append(" AA");
+        }
+        if (msg.isTruncated()) {
+            hasComma = false;
+            buf.append(" TC");
+        }
+        if (msg.isRecursionAvailable()) {
+            hasComma = false;
+            buf.append(" RA");
+        }
+        if (msg.z() != 0) {
+            if (!hasComma) {
+                buf.append(',');
+            }
+            buf.append(" Z: ");
+            buf.append(msg.z());
+        }
+
+        if (hasComma) {
+            buf.setCharAt(buf.length() - 1, ')');
+        } else {
+            buf.append(')');
+        }
+    }
+
+    private static void appendAddresses(StringBuilder buf, DnsMessage msg) {
+
+        if (!(msg instanceof AddressedEnvelope)) {
+            return;
+        }
+
+        @SuppressWarnings("unchecked")
+        AddressedEnvelope<?, SocketAddress> envelope = (AddressedEnvelope<?, SocketAddress>) msg;
+
+        SocketAddress addr = envelope.sender();
+        if (addr != null) {
+            buf.append("from: ");
```

Comment:
I wouldn't do method chaining here because indentation makes the code look ugly.